EntitySearchMixin

Ichimonji10 opened the pull request:
The first commit implements and tests class `EntitySearchMixin`. The second commit integrates the mixin with several entity classes. See the individual commit messages for more details.

The current NailGun version is 0.18.2. I'd like to release version 0.19.0 today, then include this change in 0.20.0. I am marking this PR as "DO NOT MERGE" for that reason, but am posting the PR now to collect feedback and to keep track of it.

For proof that this PR functions correctly, see <del>FIXME: Fetch information from the future</del> <ins>SatelliteQE/robottelo#2610</ins>.

sthirugn commented on nailgun/entities.py:
```diff
@@ -56,9 +58,14 @@
+# have different effects:
+#
+#     product.description = None; product.update()
+#     del product.desciption; product.update()
```
description?

Ichimonji10 replied:
Products have a description field.

sthirugn replied:
sorry I meant there is a typo

Ichimonji10 replied:
Aha! Thanks for pointing this out.

Ichimonji10 replied:
:+1: Fixed.

Ichimonji10 commented on nailgun/entity_mixins.py:
```diff
@@ -931,3 +937,325 @@ def update(self, fields=None):
+            normalized.append(attrs)
+        return normalized
+
+    def search(self, fields=None, query=None, filters=None):
```
This method contains an extensive docstring that demonstrates how to use the search mixin. (The examples are a little more verbose than necessary - I decided to opt for clarity over conciseness.) If you're just coming to the search mixin for the first time, this is a great place to start.

sthirugn commented on nailgun/entities.py:
```diff
@@ -1970,27 +1983,14 @@ def create_missing(self):
-                    'Could not find the "Library" lifecycle environment for '
-                    'organization {0}. Search results: {1}'
+                    u'Could not find the "Library" lifecycle environment for '
+                    u'organization {}. Search results: {}'
```
are we moving away from usage of using numbers ?

Ichimonji10 replied:
The numbers are only necessary for python 2.6 and older. Nailgun has never supported Python 2.6 - it makes use of conveniences like dict comprehensions that aren't available in such versions of Python. I just don't see any particular reason to use numbers.

JacobCallahan replied:
Does this break compatibility with those that use 2.6 with robottelo?

Ichimonji10 replied:
@JacobCallahan You're right, this will break 2.6 compatibility. I'll revert this change.

JacobCallahan replied:
:+1:

sthirugn commented:
ACK pending comments

Ichimonji10 commented on nailgun/entity_mixins.py:
```diff
@@ -931,3 +937,325 @@ def update(self, fields=None):
+        for a list of all organizations and then locally search through the
+        results for the organization named "foo"::
+
+            Organization().search(filters={'name': 'foo'})
```
The `filters` attribute is the only portion of this PR that I find to be really questionable:

* Someone who knows what they're doing can easily write a list comprehension that does everything `search_filter()` does and more (filtering on foreign key fields is not supported) quite trivially.
* I think it'd be more beneficial for us to focus our efforts on auto-generating good quality searches. That's a much tougher problem. Users shouldn't have to write queries all the time, which is something they have to do right now.
* I expect it'd be more useful to have something like an `expect` argument, where you can tell nailgun how many results you expect to be returned by any given query.

omaciel replied:
In this case is the filter doing a `name like 'foo'` query? Whereas `Organization(name='foo').search()` would be a `name == 'foo'`?

Ichimonji10 replied:
`filters` performs exact matching. The highlighted code above has the same effect as this:

```python
orgs = entities.Organization().search()
orgs = [org for org in orgs if org.name == 'foo']
```

This:

```python
Organization(name='foo').search()
```

...passes a search term to the server and lets the server conduct the search. The latter is vastly more efficient.

Unfortunately, `search_payload` is very naive right now. Here's what it'll do:

```python
>>> from nailgun import entities
>>> entities.Organization(name='foo').search_payload()
{'name': 'foo'}
```

And unfortunately, that search will fail. It'll fail silently, in fact - Satellite will respond to this search by returning **every single organization**.

Ichimonji10 replied:
By the way, the more I think about an `expect` argument, the more I want to add one in. Consider how cool this would be:

```python
    organization = Organization().search(
        query={'search': 'name="Systems Engineering"'},
        expect=1,
    )[0]
    location = Location().search(
        query={'search': 'name="Boston"'},
        expect=1,
    )[0]
    hostgroup = HostGroup().search(
        query={'search': 'name="RHEV 3.5"'},
        expect=1,
    )[0]
```

omaciel replied:
Love it!

rplevka replied:
:+1:

oshtaier replied:
>Unfortunately, search_payload is very naive right now. Here's what it'll do:
```
>>> from nailgun import entities
>>> entities.Organization(name='foo').search_payload()
{'name': 'foo'}
```
>And unfortunately, that search will fail. It'll fail silently, in fact - Satellite will respond to this search by returning >every single organization.

Could you please provide some examples why ```search_payload``` is not so good at that moment?
And explain more text above. That part is not obvious for me as providing organization name should be sufficient to find particular organzation

Ichimonji10 replied:
@oshtaier This will return every single organization:

    GET /katello/api/v2/organizations
    {'name': 'foo'}

This will return all organizations with a name (equaling? like? containing?) foo:

    GET /katello/api/v2/organizations
    {'search': 'name=foo'}

This will return no organizations at all:

    GET /katello/api/v2/organizations
    {'search': 'name:foo'}

The search syntax is undefined, no errors are ever returned when an invalid search is submitted, and exactly which fields can be submitted in an outer hash versus an inner "search" hash varies on a per-URL basis.

So yes, providing a name is sufficient to find a particular organization. But how do you provide that name to the API? Not in a way that you might expect.

Ichimonji10 replied:
As an aside, the practice of passing parameters in the body of a GET request is horrible. It's absolutely horrible and violates HTTP semantics. Satellite should *NOT* support it. HTTP-aware network intermediaries are at liberty to drop the body of any GET request; caches and proxies will be confused by the practice; developers, sysadmins and network admins that know a thing or two about HTTP will cringe at this.

JacobCallahan replied:
That last point is exactly the sentiment I encountered while trying to find elegant solutions for this exact issue! Not sure if there is a better way for them to implement it, unless they create a new path or something.

Ichimonji10 replied:
@JacobCallahan Aye. It's a nut of a problem. I can think of some solutions, like adding a "filter" attribute to fields. But the best thing for me to do might be to just spend some time talking with the devs and better understanding their design.

omaciel commented on nailgun/entity_mixins.py:
```diff
@@ -931,3 +937,325 @@ def update(self, fields=None):
+        The rules for generating a search query are simple, and can be
+        illustrated by example::
+
+            >>> SomeEntity(
```
I'm confused by the example here. The docstring mentions that this is an example for generating a search query... should this have used `search_payload` instead?

Ichimonji10 replied:
Yes. Good catch.

omaciel commented on nailgun/entity_mixins.py:
```diff
@@ -931,3 +937,325 @@ def update(self, fields=None):
+        * :meth:`search` may alter ``fields`` and ``query`` at will.
+        * :meth:`search_payload` may alter ``fields`` and ``query`` in an
+          idempotent manner.
+        * No other method should atler ``fields`` or ``query``.
```
s/atler/alter

Ichimonji10 replied:
:+1:

omaciel commented on nailgun/entity_mixins.py:
```diff
@@ -931,3 +937,325 @@ def update(self, fields=None):
+            results = org.search(fields={'name', 'description'})  # Same.
+            results = org.search(fields={'name'})  # Search by name.
+            results = org.search(fields={'description'})  # Search by desc
+            results = org.search(fields=set())  # Search for all organizations.
```
The syntax here sort of goes against the 'pattern' used for the other examples. Imho, if I wanted to search for all organizations, I'd call `results = Organization().search()`. Using `fields=set()` is not as intuitive... to me :) I don't have a solution or recommendation per se, but if `results = Organization().search()` == `results = Organization().search(fields=set())` or `results = Organization().search()` == `results = org.search(fields=set())` I guess I'm ok with considering the use of `fields=set()` as a 'hidden feature' :)

Ichimonji10 replied:
As of right now, all of these statements hold true:

```python
Organization().search() == Organization().search(fields=set())
Organization().search() == Organization().search(set())
Organization().search() == org.search(fields=set())
Organization().search() == org.search(set())
```

I've done this in order to make `search()` similar to `update()`. `update()` has this signature:

```python
def update(self, fields=None):
```

By default, `update()` sends all fields:

```python
org = entities.Organization(id=1, description='new description').update()
```

But you can also choose which fields to send:

```python
org = entities.Organization(id=1).read()
org.description = 'new description'
org = org.update(['description'])
```

Similarly, `search()` uses all fields:

```python
orgs = entities.Organization(description='new description').search()
```

But you may want to choose which fields to search by:

```python
org = entities.Organization(id=1).read()
orgs = org.search({'name'})
assert org.name == orgs[0].name
```

Now, you absolutely are right in pointing out that it's unusual to ask for a `set()` as an argument. That's pretty new.

Thoughts?

Ichimonji10 replied:
BTW, @elyezer has opined that it would be good to make `query` the first argument, simply because that's going to be a commonly used argument. Here's why I opted for `fields` as the first argument:

* Doing so makes `search()` similar to `update()`: the very first argument to both is `fields`.
* Doing so pushes people toward using the `fields` argument when possible. I'd like to focus our efforts on improving the `search_payload` method instead of just hacking around its naive implementation for ever and ever.

I can see the flip side, though. Thoughts?

omaciel replied:
Totally onboard with keeping argument ordering consistently. Sadly I don't have any clever suggestions to propose, but I think that once we start using this code we will figure out what the pain points are, etc. So I'd vote to proceed with the current design and see where it will take us. Also, it will be great to hear from @oshtaier @abalakh @elyezer et all for the opinion.

omaciel commented on nailgun/entity_mixins.py:
```diff
@@ -931,3 +937,325 @@ def update(self, fields=None):
+        the manual search query is melded in to the generated query. This can
+        be used to great effect::
+
+            Organization(name='foo').search(query={'per_page': 50})
```
Love it!

omaciel commented:
ACK

oshtaier commented on nailgun/entities.py:
```diff
@@ -1970,27 +1983,14 @@ def create_missing(self):
-            )
-            response.raise_for_status()
-            results = response.json()['results']
+            results = self.search({'organization'}, {u'name': u'Library'})
```
Could we use such statements here?:
```
results = self.search({u'organization_id': self.organization}, {u'name': u'Library'})
```
or
```
results = self.search({u'organization_id': self.organization.id}, {u'name': u'Library'})
```
I just don't understand difference between ```{'organization'}``` and ```{u'name': u'Library'}``` parameters representation

Ichimonji10 replied:
```python
>>> set(['organization'])
{'organization'}
>>> {'organization'}
{'organization'}
>>> {'organization'} == set(['organization'])
True
```

Ichimonji10 replied:
> Could we use such statements here?:
> ```
> results = self.search({u'organization_id': self.organization}, {u'name': u'Library'})
> ```
> or
> ```
> results = self.search({u'organization_id': self.organization.id}, {u'name': u'Library'})
> ```

No. The first parameter to `search()`, `fields`, is a set of field names. Do not pass a dict where a set is expected.

Ichimonji10 replied:
> I think about search functionality always in terms ```.search(file='something')```, so it is not so obvious for me what ```.search(file)``` should do

That's a fine thing to do if you want to write imperative code. But I'm doing my darndest to take advantage of the fact that Python is an object oriented language.

Ichimonji10 replied:
> I just don't understand difference between `{'organization'}` and `{u'name': u'Library'}` parameters representation

The first argument specifies which fields nailgun should use when automatically generating a search expression for you. The second argument is merged in to the first - it allows for customization after a search has been generated. Here's a simple example of how the queries are merged:

```python
>>> from nailgun import entities
>>> org = entities.Organization(name='foo', description='bar')                                            
>>> org.search_payload()
{'name': 'foo', 'description': 'bar'}
>>> org.search_payload({'name'})
{'name': 'foo'}
>>> org.search_payload({'name'}, {'per_page': 50})
{'per_page': 50, 'name': 'foo'}
```

I will fully admit that the search generator currently *sucks*. It's awfully naive. But I'm OK with that - I know that we have the tools in hand to make it better, and I want us to put some effort in to doing so rather than writing manual search queries for the rest of our days.

JacobCallahan replied:
:+1:

oshtaier commented on nailgun/entity_mixins.py:
```diff
@@ -931,3 +937,325 @@ def update(self, fields=None):
+
+        Make an HTTP GET call to ``self.path('base')``. Return the response.
+
+        .. WARNING:: Subclasses should not fiddle with the ``fields`` or
```
Could you please clarify more that warning message?

Ichimonji10 replied:
Let's say that the `User` class inherits from this mixin like so:

```python
class User(Entity, EntitySearchMixin):
    pass
```

In this case, `User` is a subclass of both `Entity` and `EntitySearchMixin`. That method may define a `search_raw` method:

```python
class User(Entity, EntitySearchMixin):
    def search_raw(self, fields=None, query=None):
        super(User, self).search_raw(fields, query)
```

So far, so good. But the subclass should *not* fiddle with either `fields` or `query`. This is bad:

```python
class User(Entity, EntitySearchMixin):
    def search_raw(self, fields=None, query=None):
        fields.add{'name'}
        query['per_page'] = 50
        super(User, self).search_raw(fields, query)
```

oshtaier replied:
But we have similar idea for, let's say, for read method:
```
class Domain(Entity, EntityReadMixin):

    def read(self, entity=None, attrs=None, ignore=()):
        if attrs is None:
            attrs = self.read_json()
        attrs['domain_parameters_attributes'] = attrs.pop('parameters')
        return super(Domain, self).read(entity, attrs, ignore)
```
Here, we change parameter and send it to super class

Or it is completely different?

Ichimonji10 replied:
You can fiddle with those arguments at will in `read`. From `search_normalize`:

```
At this time, it is possible to parse all search results without
knowing what search query was used. However, it is possible that
certain responses can only be parsed if the search query is known. If
that is the case, this method will be given a new ``payload`` argument,
where ``payload`` is what was sent to the server. This will only work
if ``search_raw`` and ``search`` are able to generate identical
payloads. Pending extensive use of :class:`EntitySearchMixin`:

* :meth:`search` may alter ``fields`` and ``query`` at will.
* :meth:`search_payload` may alter ``fields`` and ``query`` in an
    idempotent manner.
* No other method should alter ``fields`` or ``query``.
```

oshtaier commented on nailgun/entity_mixins.py:
```diff
@@ -931,3 +937,325 @@ def update(self, fields=None):
+            results = org.search(fields={'name'})  # Search by name.
+            results = org.search(fields={'description'})  # Search by desc
+            results = org.search(fields=set())  # Search for all organizations.
+            results = org.search(fields={'title'})  # Error!
```
Here we have error, because in ```Organization(name='foo', description='bar')``` we did not specify ```title``` field value explicitly, right?

omaciel replied:
`Organization` doesn't have a `title` field

Ichimonji10 replied:
@omaciel nailed it.

> Here we have error, because in `Organization(name='foo', description='bar')` we did not specify `title` field value explicitly, right?

Correct. You're telling nailgun to create a search using the "title" attribute, but the "title" attribute is not set.

> and at that moment it has `None` value?

Incorrect. No `title` attribute has been set at all.

```python
>>> from nailgun.entities import Organization
>>> org = Organization(name='foo', description='bar').
>>> hasattr(org, 'title')
False
>>> org.title
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Organization' object has no attribute 'title'
```

Further example:

```python
>>> from nailgun.entities import Organization
>>> org = Organization(name='foo', description='bar')                                                     
>>> org.search_payload()
{'description': 'bar', 'name': 'foo'}
>>> org.search_payload({'name'})
{'name': 'foo'}
>>> org.search_payload({'title'})                                                                         
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ichimonji10/code/nailgun/nailgun/entity_mixins.py", line 1012, in search_payload
    value = getattr(self, field)
AttributeError: 'Organization' object has no attribute 'title'
```

oshtaier replied:
Organization actually has 'title' field
```
class Organization():
...
    def __init__(self, server_config=None, **kwargs):
        self._fields = {
            'compute_resource': entity_fields.OneToManyField(
                AbstractComputeResource
            ),
            'config_template': entity_fields.OneToManyField(ConfigTemplate),
            'description': entity_fields.StringField(),
            'domain': entity_fields.OneToManyField(Domain),
            'environment': entity_fields.OneToManyField(Environment),
            'hostgroup': entity_fields.OneToManyField(HostGroup),
            'label': entity_fields.StringField(str_type='alpha'),
            'media': entity_fields.OneToManyField(Media),
            'name': entity_fields.StringField(required=True),
            'realm': entity_fields.OneToManyField(Realm),
            'smart_proxy': entity_fields.OneToManyField(SmartProxy),
            'subnet': entity_fields.OneToManyField(Subnet),
            'title': entity_fields.StringField(),
            'user': entity_fields.OneToManyField(User),
        }
```
or API call
```
"results": [{"id":5,"name":"𩀵𨡼𥲲𨐿䎘𦮱","title":"𩀵𨡼𥲲𨐿䎘𦮱","label":"547374fa-2a20-4c14-8ba2-9fa099e6a396","created_at":"2015-07-20T10:09:43Z", ........
```

So we can use only name and description in your example as search criteria, right?

Ichimonji10 replied:
> Organization actually has 'title' field

Correct.

But merely having a field is useless when generating a search query. In order to generate a search query, some attributes must be present.

I don't understand your questions completely, but I may be able to answer with another example. Let's take the object you've shown:

```python
>>> from nailgun.config import ServerConfig
>>> cfg = ServerConfig('https://hp-ml310egen8-01.rhts.eng.bos.redhat.com', auth=('admin', 'changeme'), verify=False)
>>> cfg.save('oshtaier')
>>> cfg = ServerConfig.get('oshtaier')
>>> org = entities.Organization(cfg, id=5).read()
>>> org.search_payload().keys()
dict_keys(['title', 'config_template_ids', 'domain_ids', 'compute_resource_ids', 'hostgroup_ids', 'environment_ids', 'user_ids', 'smart_proxy_ids', 'id', 'subnet_ids', 'label', 'media_ids', 'description', 'name'])
```

See how all of these different keys are included in the search?

JacobCallahan replied:
@oshtaier be careful. the line ``org = Organization(name='foo', description='bar')``` has only created an Organization locally. At this point, the Org only has those two fields set. This will remain the case until the Org is assigned a title or is created at the server. See below for some examples.

```
In [5]: org = entities.Organization(name='foo', description='bar')

In [6]: org
Out[6]: nailgun.entities.Organization(nailgun.config.ServerConfig(url=u'https://rhsm-qe-1.rhq.lab.eng.bos.redhat.com/', auth=(u'admin', u'changeme')), description='bar', name='foo')

In [7]: org.get_values()
Out[7]: {'description': 'bar', 'name': 'foo'}

In [8]: org.get_fields()
Out[8]: 
{'compute_resource': <nailgun.entity_fields.OneToManyField at 0x7ff314bf79d0>,
 'config_template': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7a10>,
 'description': <nailgun.entity_fields.StringField at 0x7ff314bf7a50>,
 'domain': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7a90>,
 'environment': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7ad0>,
 'hostgroup': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7b10>,
 'id': <nailgun.entity_fields.IntegerField at 0x7ff314bf7e10>,
 'label': <nailgun.entity_fields.StringField at 0x7ff314bf7b50>,
 'media': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7b90>,
 'name': <nailgun.entity_fields.StringField at 0x7ff314bf7bd0>,
 'realm': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7c10>,
 'smart_proxy': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7c50>,
 'subnet': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7c90>,
 'title': <nailgun.entity_fields.StringField at 0x7ff314bf7cd0>,
 'user': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7d10>}

In [9]: org.title
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-9-5f07d6148295> in <module>()
----> 1 org.title

AttributeError: 'Organization' object has no attribute 'title'
```

JacobCallahan commented:
@Ichimonji10 I guess my single biggest concern is that the default results per page is 20. This could result in a number of failed searches, especially when dealing with automation. For example, over the lifecycle of the machines we currently run our automation on, there are Organizations with id's exceeding 530! Of course, a portion of these are deleted but most remain. For most non-automated users, they may never encounter this problem. But if we are running automation and are searching for an organization we just created, there is a high probability that the search will fail.

There are two main solutions to this issue. 
1. Hand it off to the person running the search. This would basically mean that we need to include a large page count in each search query. That is fine, but maybe not optimal.
2. Nailgun defaults to a per_page that is large enough to return a significant number of results. This helps to ensure our target is included in the results, but can quickly bog down a nailgun-based search. 

Of course, it would be optimal if the person performing the search includes a server-side filter. Just my thoughts.

JacobCallahan commented:
This has come a long way from our initial prototypes. Good job!
ACK, pending the few remaining comments and @oshtaier.

Ichimonji10 commented:
@JacobCallahan 

Whether a search returns a lot of results depends on what we're keying our search on, right? For example, if I'm searching for something by name or label, then only one result should be returned - no matter how many organizations exist.

omaciel commented:
> 1. Hand it off to the person running the search. This would basically mean that we need to include a large page count in each search query. That is fine, but maybe not optimal.

@JacobCallahan I think that if someone writing a test specifically wants all entries for $ENTITY, it would be reasonable that he/she passes a filter for how many results to fetch (aka `per_page`)... but, perhaps the search method in a future version should return a generator instead of a list of results? That may work as data would be fetched lazily?

JacobCallahan commented:
@Ichimonji10 Yep, that is why I included that last line. However, if someone were to run something like this on our automation boxes, they would be sorely disappointed. 
```my_org = Organization().search(filters={'name': 'just_created'})```

@omaciel Yes, that is what I was thinking for my first solution option.

Ichimonji10 commented:
@Ichimonji10 Yep, that is why I included that last line. However, if someone were to run something like this on our automation boxes, they would be sorely disappointed. 
`my_org = Organization().search(filters={'name': 'just_created'})`

@JacobCallahan Believe it or not...

```python
>>> from nailgun import entities
>>> orgs = entities.Organization().search()
>>> len(orgs)
627
```

Ichimonji10 commented:
@JacobCallahan Anyway, yes, a user would be disappointed. But I think that we can reasonably file a bug against the server for that issue, where the bug says "invalid searches are silently ignored."

oshtaier commented:
ACK

Ichimonji10 commented:
Hey all! I've taken your feedback in to account and tweaked the pull request. As it stands, I've made only very minor tweaks to the code itself. The majority of my changes have to do with documentation. I've added a short section in to the "examples" document and fiddled with many of the docstrings. There are three changes of particular note:

First, I've made changes to the "Examples" document.

Second, I've expanded the docstring for method `EntitySearchMixin.search_payload` and added several links pointing to it. It now gives more insight in to what the final search queries look li.e

Third, I'm now referencing `LifecycleEnvironment` in my examples instead of `Organization`. Why? Because this is an invalid search:

    GET /katello/api/v2/organizations
    {'name': 'foo', 'description': 'bar'}  # server ignores these params

But this is a valid search:

    GET /katello/api/v2/lifecycle_environments
    {'name': 'foo', 'organization_id': 1}  # server accepts these params

The default implementation of `search_payload` is pretty naive. But I at least want users to be able to copy and paste the examples in the documentation!

I'm currently running `make test-foreman-api-threaded` (Thanks @JacobCallahan!) as a final santiy check before opening this up for merging.

omaciel commented:
w00t!

Ichimonji10 commented:
I've done a sanity check here: https://github.com/SatelliteQE/robottelo/pull/2610#issuecomment-124696030

I think this PR is open for merging.

oshtaier commented:
Re-ACK. Thanks @Ichimonji10 for you patience
:+1: For change from ```Organization``` entity to ```LifecycleEnvironment```

Ichimonji10 commented:
:tada:

elyezer commented:
:+1:
